File: app/src/main/java/com/antyzero/smoksmog/ui/screen/start/StationSlideAdapter.java
package com.antyzero.smoksmog.ui.screen.start;

import android.app.FragmentManager;
import android.support.v13.app.FragmentStatePagerAdapter;
import android.support.v4.view.PagerAdapter;
import android.util.SparseArray;
import android.view.ViewGroup;

import com.antyzero.smoksmog.ui.screen.start.fragment.StationFragment;

import java.lang.ref.WeakReference;
import java.util.List;

/**
 * Adapter for sliding pages left-right
 */
public class StationSlideAdapter extends FragmentStatePagerAdapter {

    private final List<Long> stationIds;

    private SparseArray<WeakReference<StationFragment>> fragmentRegister = new SparseArray<>();

    /**
     * Standard constructor
     *
     * @param fragmentManager for fragment manipulations
     * @param stationIds      list of IDs for data download
     */
    public StationSlideAdapter(FragmentManager fragmentManager, List<Long> stationIds) {
        super(fragmentManager);
        this.stationIds = stationIds;
    }

    @Override
    public int getItemPosition(Object object) {
        return PagerAdapter.POSITION_NONE;
    }

    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        StationFragment fragment = (StationFragment) super.instantiateItem(container, position);
        fragmentRegister.put(position, new WeakReference<>(fragment));
        return fragment;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        fragmentRegister.remove(position);
        super.destroyItem(container, position, object);
    }

    @Override
    public StationFragment getItem(int position) {
        return StationFragment.newInstance(stationIds.get(position));
    }

    public WeakReference<StationFragment> getFragmentReference(int position) {
        return fragmentRegister.get(position);
    }

    @Override
    public int getCount() {
        return stationIds.size();
    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/ui/screen/start/StationSlideAdapter.kt
package com.antyzero.smoksmog.ui.screen.start

import android.app.FragmentManager
import android.support.v13.app.FragmentStatePagerAdapter
import android.support.v4.view.PagerAdapter
import android.util.SparseArray
import android.view.ViewGroup
import com.antyzero.smoksmog.storage.model.Item

import com.antyzero.smoksmog.ui.screen.start.fragment.StationFragment

import java.lang.ref.WeakReference

/**
 * Adapter for sliding pages left-right
 */
class StationSlideAdapter(fragmentManager: FragmentManager, private val stationIds: List<Item>) : FragmentStatePagerAdapter(fragmentManager) {

    private val fragmentRegister = SparseArray<WeakReference<StationFragment>?>()

    override fun getItemPosition(`object`: Any?): Int {
        return PagerAdapter.POSITION_NONE
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val fragment = super.instantiateItem(container, position) as StationFragment
        fragmentRegister.put(position, WeakReference(fragment))
        return fragment
    }

    override fun destroyItem(container: ViewGroup?, position: Int, `object`: Any) {
        fragmentRegister.remove(position)
        super.destroyItem(container, position, `object`)
    }

    override fun getItem(position: Int): StationFragment {
        return StationFragment.newInstance(stationIds[position].id)
    }

    fun getFragmentReference(position: Int): WeakReference<StationFragment>? {
        return fragmentRegister.get(position)
    }

    override fun getCount(): Int {
        return stationIds.size
    }
}
